tests_e2e/scenarios/lib/vm_extension.py

```diff
@@ -0,0 +1,239 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2018 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#
+# This module includes facilities to execute VM extension operations (enable, remove, etc) on single virtual machines (using
+# class VmExtension) or virtual machine scale sets (using class VmssExtension).
+#
+
+import uuid
+
+from abc import ABC, abstractmethod
+from assertpy import assert_that, soft_assertions
+from typing import Any, Callable, Dict, Type
+
+from azure.core.polling import LROPoller
+from azure.mgmt.compute import ComputeManagementClient
+from azure.mgmt.compute.models import VirtualMachineExtension, VirtualMachineScaleSetExtension, VirtualMachineExtensionInstanceView
+from azure.identity import DefaultAzureCredential
+
+from tests_e2e.scenarios.lib.identifiers import VmIdentifier, VmExtensionIdentifier
+from tests_e2e.scenarios.lib.logging import log
+from tests_e2e.scenarios.lib.retry import execute_with_retry
+
+
+_TIMEOUT = 5 * 60  # Timeout for extension operations (in seconds)
+
+
+class _VmExtensionBaseClass(ABC):
+    """
+    Abstract base class for VmExtension and VmssExtension.
+
+    Implements the operations that are common to virtual machines and scale sets. Derived classes must provide the specific types and methods for the
+    virtual machine or scale set.
+    """
+    def __init__(self, vm: VmIdentifier, extension: VmExtensionIdentifier, resource_name: str):
+        super().__init__()
+        self._vm: VmIdentifier = vm
+        self._identifier = extension
+        self._resource_name = resource_name
+        self._compute_client: ComputeManagementClient = ComputeManagementClient(credential=DefaultAzureCredential(), subscription_id=vm.subscription)
+
+    def enable(
+        self,
+        settings: Dict[str, Any] = None,
+        protected_settings: Dict[str, Any] = None,
+        auto_upgrade_minor_version: bool = True,
+        force_update: bool = False,
+        force_update_tag: str = None
+    ) -> None:
+        """
+        Performs an enable operation on the extension.
+
+        NOTE: 'force_update' is not a parameter of the actual ARM API. It is provided for convenience: If set to True,
+              the 'force_update_tag' can be left unspecified and this method will generate a random tag.
```

Comment:
What is the force_update_tag used for?

Reply:
If the extension settings do not change, invoking the extension won't generate a new goal state (because the model has not changed). To work around that, CRP exposes the force_update_tag, which is an arbitrary string with no meaning other than providing a way to specify unique values on different calls to enable (our code uses GUIDs)